Android/app/src/go/doh/doh_test.go

```diff
@@ -889,3 +863,77 @@ func TestServfail(t *testing.T) {
 		t.Errorf("Wrong question: %v", servfail.Questions[0])
 	}
 }
+
+func TestQueryCanBeCancelled(t *testing.T) {
+	expectDoHTimeout := func(server testingDoHServer, msg string) {
+		doh := makeTestDoHTransport(t, server)
+		st := time.Now()
+		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+		defer cancel()
+		doh.Query(ctx, simpleQueryBytes)
+		require.WithinRange(t, time.Now(), st.Add(1700*time.Millisecond), st.Add(2300*time.Millisecond), msg)
+	}
+
+	expectDoHTimeout(unreachableDoH, "unreachable server should timeout within deadline of ctx")
+
+	// Intentionally create a local DoH server that does not accept any requests
+	addr, err := url.Parse(localDoH.url)
+	require.NoError(t, err)
+	dohAddr, err := net.ResolveTCPAddr("tcp", addr.Host)
+	require.NoError(t, err)
+	svr, err := net.ListenTCP("tcp", dohAddr)
+	require.NoError(t, err)
+	defer svr.Close()
+
+	expectDoHTimeout(localDoH, "unresponsive server should timeout within deadline of ctx")
+}
+
+/******** Test DoH servers ********/
+type testingDoHServer struct {
+	url string
+	ips []string
+}
+
+var unreachableDoH = testingDoHServer{
+	url: "https://1.2.3.4:443",
+	ips: []string{"1.2.3.4"},
+}
+
+var googleDoH = testingDoHServer{
+	url: "https://dns.google/dns-query",
+	ips: []string{
+		"8.8.8.8",
+		"8.8.4.4",
+		"2001:4860:4860::8888",
+		"2001:4860:4860::8844",
+	},
+}
+
+var localDoH = testingDoHServer{
+	url: "https://localhost:34443",
+	ips: []string{"127.0.0.1"},
+}
+
+/********** DoH Transport Test Helpers **********/
+type testingDoHListener struct {
+	Listener
+	summary *Summary
+}
+
+func (l *testingDoHListener) OnQuery(url string) Token         { return nil }
+func (l *testingDoHListener) OnResponse(tok Token, s *Summary) { l.summary = s }
+
+func makeTestDoHTransport(t *testing.T, target testingDoHServer) Transport {
+	doh, err := NewTransport(target.url, target.ips, nil, nil, nil)
+	require.NoError(t, err)
+	require.NotNil(t, doh)
+	return doh
+}
```

Comment:
Something like this would be much less confusing:
```suggestion
func newTestDOHResolver(t *testing.T, config dohConfig) DOHResolver {
	doh, err := NewDOHResolver(target.url, target.ips, nil, nil, nil)
	require.NoError(t, err)
	require.NotNil(t, doh)
	return doh
}
```

Reply:
BTW Go tends to prefer "new" over "make".